crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -172,92 +173,131 @@ impl<'w, E, B: Bundle> DerefMut for Trigger<'w, E, B> {
 /// will run.
 pub trait TriggerTargets {
     /// The components the trigger should target.
-    fn components(&self) -> &[ComponentId];
+    fn components(&self) -> impl ExactSizeIterator<Item = ComponentId> + Clone;
```

Comment:
```suggestion
    fn components(&self) -> impl ExactSizeIterator<Item = ComponentId> + Clone + '_;
```
Could we connect the return type to the `TriggerTargets`' lifetime to avoid allocations in the `Vec<T>` impl?

Reply:
The question isn't resolved because the `collect`s are still present, and it's not done for `TriggerTargets::entities` either.

Reply:
@ItsDoot Now it is the wrong kind of iterator. Should I change the return type?

Reply:
I just tested it locally, you should be able to swap `ExactSizeIterator` to just `Iterator`

Reply:
Done